src/probnum/filtsmooth/gaussfiltsmooth/kalman_utils.py

```diff
@@ -0,0 +1,215 @@
+
+
+def _iterated_update(
+    update_fun, stopcrit, measurement_model, rv, time, data, _linearise_at=None
+):
+    """Turn an update_*() function into an iterated update.
+
+    This iteration is continued until it reaches a fixed-point (as measured with atol and rtol).
+    Using this inside `Kalman` yields the iterated (extended/unscented/...) Kalman filter.
+
+    Examples
+    --------
+    >>> import functools as ft
+    >>>
+    >>> stopcrit = StoppingCriterion(atol=1e-2, rtol=1e-4, maxit=1000)
+    >>> iterated_update_classic = ft.partial(iterated_update, update_fun=update_classic, stopcrit=stopcrit)
+    """
+    current_rv, meas_rv, info = update_fun(
+        measurement_model=measurement_model,
+        rv=rv,
+        time=time,
+        data=data,
+        _linearise_at=_linearise_at,
+    )
+    new_mean = current_rv.mean
+    old_mean = np.inf * np.ones(current_rv.mean.shape)
+    while not stopcrit.terminate(error=new_mean - old_mean, reference=new_mean):
+        old_mean = new_mean
+        current_rv, meas_rv, info = update_fun(
+            measurement_model=measurement_model,
+            rv=current_rv,
+            time=time,
+            data=data,
+        )
+        new_mean = current_rv.mean
+    return current_rv, meas_rv, info
+
+
+########################################################################################################################
+# Smoothing choices
+########################################################################################################################
+
+
+def rts_with_precon(smooth_step_fun):
+    """Make a smoothing step respect preconditioning.
+
+    Currently, this is only available for IBM() integrators.
+
+    Examples
+    --------
+    >>> import functools as ft
+    >>> rts_smooth_step_classic_with_precon = rts_with_precon(rts_smooth_step_classic)
+    """
+
+    def new_smoothing_function(*args, **kwargs):
+        return _rts_smooth_step_with_precon(smooth_step_fun, *args, **kwargs)
+
+    return new_smoothing_function
+
+
+def _rts_smooth_step_with_precon(
+    smooth_step_fun,
+    unsmoothed_rv,
+    predicted_rv,
+    smoothed_rv,
+    crosscov,
+    dynamics_model=None,
+    start=None,
+    stop=None,
+):
+    """Execute a smoothing step with preconditioning."""
+
+    # Assemble preconditioners
+    dt = stop - start
+    precon = dynamics_model.precon(dt)
+    precon_inv = dynamics_model.precon.inverse(dt)
+
+    # Pull RVs/matrices into preconditioned space
+    unsmoothed_rv = precon_inv @ unsmoothed_rv
+    predicted_rv = precon_inv @ predicted_rv
+    smoothed_rv = precon_inv @ smoothed_rv
+    crosscov = precon_inv @ crosscov @ precon_inv.T
+
+    # Carry out the smoothing step
+    updated_rv, _ = smooth_step_fun(
+        unsmoothed_rv,
+        predicted_rv,
+        smoothed_rv,
+        crosscov,
+        dynamics_model=dynamics_model,
+        start=start,
+        stop=stop,
+    )
+    new_rv = precon @ updated_rv
+    return new_rv, {}
```

Comment:
this structure now is not only applicable for IBM, but for any `Integrator`. IOUP and Matern use it as well now.